advanced-data-types.md

```diff
@@ -0,0 +1,195 @@
+    Error
+}
+
+class Transaction : RealmObject
+{
+    private string _State { get; set; }
+
+    public TransactionState State
+    {
+        get => Enum.Parse<TransactionState>(_State);
+        set => _State = value.ToString();
+    }
+}
+```
+
+Using this technique, you can persist many other types, as long as they can be converted losslessly to a persistable primitive type. In this case, we chose `string`, but we could have just as easily used integer. The string representation takes a bit more memory, but is also more explicit and less error prone - e.g. if you rearrange the enum members, the data will still be consistent.
+
+All that is pretty cool, but we can take it up a notch. By building on top of this idea, we can also devise a strategy for representing complex data types - such as `Vector3` in a Unity game or a `GeoCoordinate` in a location-aware app. To do so, we'll take advantage of embedded objects - a Realm concept that represents a complex data structure that is owned entirely by its parent. Embedded objects are a great fit for this use case because we want to have a strict 1:1 relationship and we want to make sure that deleting the parent also cleans up the embedded objects it owns. Let's see this in action:
+
+```csharp
+class Vector3Model : EmbeddedObject
+{
+    // Casing of the properties here is unusual for C#,
+    // but consistent with the Unity casing.
+    private float x { get; set; }
+    private float y { get; set; }
+    private float z { get; set; }
+
+    public Vector3Model(Vector3 vector)
+    {
+        x = vector.x;
+        y = vector.y;
+        z = vector.z;
+    }
+
+    private Vector3Model()
+    {
+    }
+
+    public Vector3 ToVector3() => new Vector3(x, y, z);
+}
+
+class Powerup : RealmObject
+{
+    [MapTo("Position")]
+    private Vector3Model _Position { get; set; }
+
+    public Vector3 Position
+    {
+        get => _Position?.ToVector3() ?? Vector3.Zero;
+        set => _Position = new Vector3Model(value);
+    }
+
+    protected override void OnPropertyChanged(string propertyName)
+    {
+        base.OnPropertyChanged(propertyName);
+
+        if (propertyName == nameof(_Position))
+        {
+            RaisePropertyChanged(nameof(Position));
+        }
+    }
+}
+```
+
+In this example we've defined a `Vector3Model` that roughly mirrors Unity's `Vector3` - it has 3 float properties representing the 3 components of the vector. We've also utilized what we learned in the previous sections - it has a private constructor to force consumers to always construct it with a `Vector3` argument. We've also marked its properties as private as we don't want consumers directly interacting with them - we want users to always call `ToVector3` to obtain the Unity type. And for our `Powerup` model, we're doing exactly that in the publicly exposed `Position` property. Note that similarly, to our `Person` example, we're making sure to raise a notification for `Position` whenever `_Position` changes.
```

Comment:
This actually brought up another interesting question: A `Vector3` (https://docs.unity3d.com/ScriptReference/Vector3.html) offers more than we implemented in the other tutorial and here. Would we want to show examples on how to achieve a 'full' `Vector3` that is also persistable? Do we even have a good recommendation other than waiting for polymorphic inheritance?

Reply:
The additional properties are computed values and are derived from the x, y, z coordinates. Obviously, we can't use them in queries, but similarly to the point above - this post is intended to be merely an inspiration and not so much a definitive guide. If people need to query by magnitude, they can, of course, store that as well, but it's not required to recreate the vector. We can't support the vector algebra functions in queries because Realm 1) doesn't have vector algebra functions and 2) isn't aware that this object being stored is a vector. But in the future we may be able to offer a way for users to use their own methods in queries or even implement some common vector algebra functions in Core for use in gaming scenarios.

Reply:
👍